splice/web/api/content_upload.py

```diff
@@ -0,0 +1,209 @@
+        sign_payload = []
+        for s in sign_list:
+            if s not in all_assets:
+                raise Exception("File %s in manifest is not in the content" % s)
+
+            # the content payload reference: https://github.com/mozilla-services/autograph
+            hash = hashlib.sha384(zf.read(s)).hexdigest()
+            payload = {
+                "template": "content-signature",
+                "input": "%s" % base64.b64encode(hash)
+            }
+            sign_payload.append(payload)
+
+        # send to signing server
+        signatures = dict(zip(sign_list, _sign_content(sign_payload)))
+
+        for name in all_assets:
+            yield name, zf.read(name), signatures.get(name)
+
+
+def _bump_needed(content):
+    with zipfile.ZipFile(content, 'r') as zf:
+        # parse the manifest
+        try:
+            manifest = json.loads(zf.read("manifest.json"))
+        except:
+            return False
+        else:
+            return manifest['bump_version']
+
+
+def _sign_content(sign_payload):
+    content_type = 'application/json'
+    sender = Sender({'id': 'alice',
+                     'key': env.config.SIGNING_HAWK_KEY,
+                     'algorithm': 'sha256'},         # credentials
+                    env.config.SIGNING_SERVICE_URL,  # url
+                    'POST',                          # method
+                    content=json.dumps(sign_payload),
+                    content_type=content_type)
+    try:
+        r = requests.post(env.config.SIGNING_SERVICE_URL,
+                          json=sign_payload,
+                          headers={'Content-Type': content_type,
+                                   'Authorization': sender.request_header})
+        if r.status_code != 201:
+            msg = "signing server returned an error code: %s" % (r.status_code)
+            raise Exception(msg)
+    except Exception as e:
+        msg = "Error when signing: %s" % e
+        env.log(msg)
+        raise
+    else:
+        return r.json()
+
+
+def _extract_entryption_info(signature_dict):
+    """Extract key and signature from the signing payload
+
+    Note:
+    * signing server may not return this key
+    * signing server returns multiple signatures in several encoding forms,
+      Splice only uses the first one
+    """
+    encrypt_key = signature_dict["certificate"].get("encryptionkey")
+    signature = signature_dict["signatures"][0]["signature"]
+    return encrypt_key, signature
+
+
+def upload_content_to_s3(name, version, asset, bucket, headers):  # pragma: no cover
+    headers = dict(headers)  # clone the headers as it's mutable
+    asset_name, asset_body, signature = asset
+    s3_key = "{0}/v{1}/{2}".format(name, version, asset_name)
+
+    key = Key(bucket)
+    key.name = s3_key
+    _, ext = os.path.splitext(asset_name)
+    headers['Content-Type'] = MIME_EXTENSIONS.get(ext) or 'text/plain'
```

Comment:
This is good.

We can't use libmagic here, because libmagic returns the wrong MIME type for xpi and css.